A discussion on helping downstream packaging

mgorny opened the pull request:
A discussion aimed at explaining how downstream packaging works, what challenges are the packagers facing and how they can be helped.

<!-- readthedocs-preview python-packaging-user-guide start -->
----
📚 Documentation preview 📚: https://python-packaging-user-guide--1791.org.readthedocs.build/en/1791/

<!-- readthedocs-preview python-packaging-user-guide end -->

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+- Static linking and vendoring obscures the use of external dependencies,
+  **making source auditing harder**.
+
+- Dynamic linking makes it possible to easily and **quickly replace the used
+  libraries**, which can be particularly important when they turn out to
+  be vulnerable or buggy.
```
```suggestion
- Dynamic linking makes it possible to quickly and systematically replace the used
  libraries across an entire downstream packaging ecosystem, which can be particularly
  important when they turn out to contain a security vulnerability or critical bug.
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  libraries**, which can be particularly important when they turn out to
+  be vulnerable or buggy.
+
+- Using system dependencies makes the package benefit from **downstream
+  customization** that can improve the user experience on a particular platform,
+  without the downstream maintainers having to consistently patch
+  the dependencies vendored in different packages. This can include
+  **compatibility improvements and security hardening**.
```
```suggestion
- Using system dependencies makes the package benefit from downstream
  customization that can improve the user experience on a particular platform,
  without the downstream maintainers having to consistently patch
  the dependencies vendored in different packages. This can include
  compatibility improvements and security hardening.
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  the dependencies vendored in different packages. This can include
+  **compatibility improvements and security hardening**.
+
+- Static linking and vendoring could result in **multiple different versions
+  of the same library being loaded in the same process** (e.g. when you use two
+  Python packages that link to different versions of the same library).
+  This can cause no problems, but it could also lead to anything from subtle
+  bugs to catastrophic failures.
```
```suggestion
- Static linking and vendoring can result in multiple different versions of the
  same library being loaded in the same process (for example, attempting to
  import two Python packages that link to different versions of the same library).
  This sometimes works without incident, but it can also lead to anything from library
  loading errors, to subtle runtime bugs, to catastrophic system failures.
```

mgorny replied:
Not sure if "system failures" won't be misleading here. My point was more like risk of data loss.

ncoghlan replied:
Fair, something like `... catastrophic failures (like losing data, or rendering the system unable to boot)` could be clearer.

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  This can cause no problems, but it could also lead to anything from subtle
+  bugs to catastrophic failures.
+
+- Last but not least, static linking and vendoring results in duplication,
+  and may increase the **use of both the disk space and memory**.
```
```suggestion
- Last but not least, static linking and vendoring results in duplication,
  and may increase the use of both disk space and memory.
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  and may increase the **use of both the disk space and memory**.
+
+How?
+~~~~
```
```suggestion
~~~~

```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+
+How?
+~~~~
+A good compromise between the needs of both parties is to **provide a switch
+between using vendored and system dependencies**. Ideally, if the package has
```
```suggestion
A good compromise between the needs of both parties is to provide a switch
between using vendored and system dependencies. Ideally, if the package has
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+for each dependency, and a general switch to control the default for them,
+e.g. via a ``USE_SYSTEM_DEPS`` environment variable.
+
+If the user requests using system dependencies, and **a particular dependency
+is either missing or incompatible, the build should fail** with an explanatory
```
```suggestion
If the user requests using system dependencies, and a particular dependency
is either missing or incompatible, the build should fail with an explanatory
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+message rather than fall back to a vendored version. This gives the packager
+the opportunity to notice their mistake and a chance to consciously decide
+how to solve it.
+
```
```suggestion

Note that it is reasonable for upstream projects to leave *testing* of building with
system dependencies to their downstream repackagers. The goal of these guidelines
is to facilitate more effective collaboration between upstream projects and downstream
repackagers, not to suggest upstream projects take on tasks that downstream repackagers
are better equipped to handle.
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+how to solve it.
+
+
+.. _Support downstream testing:
```
```suggestion
.. _support-downstream-testing:
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+.. _Support downstream testing:
+
+Support downstream testing
+--------------------------
```
```suggestion
--------------------------

```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+Support downstream testing
+--------------------------
+Why?
+~~~~
```
```suggestion
~~~~

```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+smoke testing to comprehensive runs of the complete test suite. There can
+be various reasons for doing this, for example:
+
+- Verifying that the downstream **packaging did not introduce any bugs**.
```
```suggestion
- Verifying that the downstream packaging did not introduce any bugs.
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+
+- Verifying that the downstream **packaging did not introduce any bugs**.
+
+- Testing on **additional platforms** that are not covered by upstream testing.
```
```suggestion
- Testing on additional platforms that are not covered by upstream testing.
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+
+- Testing on **additional platforms** that are not covered by upstream testing.
+
+- Finding subtle bugs that can only be reproduced on a **particular hardware,
+  system package versions**, and so on.
```
```suggestion
- Finding subtle bugs that can only be reproduced with particular hardware,
  system package versions, and so on.
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+- Finding subtle bugs that can only be reproduced on a **particular hardware,
+  system package versions**, and so on.
+
+- Testing the released package against **newer dependency versions** than
+  the ones present during upstream release testing.
```
```suggestion
- Testing the released package against newer (or older) dependency versions than
  the ones present during upstream release testing.
```

mgorny replied:
I'm okay with changing this, but my idea was to avoid "why aren't you using up-to-date packages?" argument.

ncoghlan replied:
Reasonable concern, but I think prioritising the "newer" case in the text will cover it (maybe that's overly optimistic, but we can always drop the parenthetical bit later if we need to)

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+- Testing the released package against **newer dependency versions** than
+  the ones present during upstream release testing.
+
+- Testing the package in an environment closely resembling **the production
+  setup**. This can detect issues caused by nontrivial interactions between
```
```suggestion
- Testing the package in an environment closely resembling the production
  setup. This can detect issues caused by nontrivial interactions between
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  different installed packages, including packages that are not dependencies
+  of your package, but nevertheless can cause issues.
+
+- Testing the released package against **newer Python versions** (including
```
```suggestion
- Testing the released package against newer Python versions (including
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+- Testing the released package against **newer Python versions** (including
+  newer point releases), or less tested Python implementations such as PyPy.
+
+Admittedly, sometimes downstream testing may yield false positives or
+inconvenience you about scenarios that you are not interested in supporting.
```
```suggestion
Admittedly, sometimes downstream testing may yield false positives or bug
reports about scenarios the upstream project is not interested in supporting.
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+
+Admittedly, sometimes downstream testing may yield false positives or
+inconvenience you about scenarios that you are not interested in supporting.
+However, perhaps even more often it does provide early notice of problems,
+or find nontrivial bugs that would otherwise cause issues for your users
+in production. And believe me, the majority of **downstream packagers are doing
+their best to double-check their results, and help you triage and fix the bugs
+that they report**.
```
```suggestion
However, perhaps even more often it does provide early notice of problems,
or find non-trivial bugs that would otherwise cause issues for the upstream
project's users. While mistakes do happen, the majority of downstream packagers
are doing their best to double-check their results, and help their upstream projects
triage and fix the bugs that they report.
```

mgorny replied:
I think "their" is mixing subjects here at the end, or at least "downstream packagers' upstream projects" sounds weird to me.

ncoghlan replied:
Yeah, the wording you used instead is better than what I proposed.

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+that they report**.
+
+How?
+~~~~
```
```suggestion
~~~~

```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+better. Some of them were already mentioned in this discussion. Some examples
+are:
+
+- **Include the test files and fixtures in the source distribution**, or make it
+  possible to easily download them separately.
```
```suggestion
- Include the test files and fixtures in the source distribution, or make it
  possible to easily download them separately.
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+
+How?
+~~~~
+There is a number of things that you can do to help us test your package
+better. Some of them were already mentioned in this discussion. Some examples
+are:
```
```suggestion
There are a number of things that upstream projects can do to help downstream
repackagers test their packages efficiently and effectively, including some of the suggestions
already mentioned above. These are typically improvements that make the test suite more
reliable and easier to use for everyone, not just downstream packagers.
Some specific suggestions are:
```

mgorny replied:
That's very nice, thanks!

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+- **Include the test files and fixtures in the source distribution**, or make it
+  possible to easily download them separately.
+
+- **Do not write to the package directories during testing.** Downstream test
```
```suggestion
- Do not write to the package directories during testing. Downstream test
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  performed during testing and temporary test files may end up being part
+  of the installed package!
+
+- **Make the test suite work offline.** Mock network interactions, using
```
```suggestion
- Make the test suite work offline. Mock network interactions, using
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+
+- **Make the test suite work offline.** Mock network interactions, using
+  packages such as responses_ or vcrpy_. If that is not possible, make it
+  possible to easily disable the tests using Internet access, e.g. via a pytest
```
```suggestion
  possible to easily disable the tests using Internet access, e.g. via a pytest_
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  marker.  Use pytest-socket_ to verify that your tests work offline. This
+  often makes your own test workflows faster and more reliable as well.
+
+- **Make your tests work without a specialized setup**, or perform the necessary
```
```suggestion
- Make your tests work without a specialized setup, or perform the necessary
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  to system services such as databases — in an extreme case, you could crash
+  a production service!
+
+- **If your package has optional dependencies, make their tests optional as
+  well.** Either skip them if the needed packages are not installed, or add
```
```suggestion
- If your package has optional dependencies, make their tests optional as
  well. Either skip them if the needed packages are not installed, or add
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  well.** Either skip them if the needed packages are not installed, or add
+  markers to make deselecting easy.
+
+- More generally, **add markers to tests with special requirements**. These can
```
```suggestion
- More generally, add markers to tests with special requirements. These can
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  include e.g. significant space usage, significant memory usage, long runtime,
+  incompatibility with parallel testing.
+
+- **Do not assume that the test suite will be run with -Werror.** Downstreams
```
```suggestion
- Do not assume that the test suite will be run with ``-Werror``. Downstreams
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  dependency versions. Assert for warnings using ``pytest.warns()`` rather
+  than ``pytest.raises()``!
+
+- **Aim to make your test suite reliable and reproducible.** Avoid flaky tests.
```
```suggestion
- Aim to make your test suite reliable and reproducible. Avoid flaky tests.
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  has its advantages, but you want to have static test cases for completeness
+  as well.
+
+- **Split tests by their purpose, and make it easy to skip categories that are
+  irrelevant or problematic.** Since the primary purpose of downstream testing
```
```suggestion
- Split tests by their purpose, and make it easy to skip categories that are
  irrelevant or problematic. Since the primary purpose of downstream testing
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+
+- **Split tests by their purpose, and make it easy to skip categories that are
+  irrelevant or problematic.** Since the primary purpose of downstream testing
+  is to ensure that the package itself works, we generally are not interested
+  in e.g. checking code coverage, code formatting, typing or running
+  benchmarks. These tests can fail as dependencies are upgraded or the system
+  is under load, without actually affecting the package itself.
```
```suggestion
  is to ensure that the package itself works, downstreams are not generally interested
  in tasks such as checking code coverage, code formatting, typechecking or running
  benchmarks. These tests can fail as dependencies are upgraded or the system
  is under load, without actually affecting the package itself.
```

ncoghlan replied:
I left the examples of skipped checks as is in my edit, but skipping typechecking when upgrading dependencies actually seems a bit dubious - typecheckers will complain if an upgrade removes or changes interfaces in a dependency that a project accesses from lines that the test suite doesn't cover.

mgorny replied:
Oh, I see that you've clarified here it's about uncovered tests. I've already replied to the top-level comment.

ncoghlan replied:
Good point. Since "to typecheck, or not to typecheck" is a decision downstream packagers will want to make on a case by case basis, it's a valid entry in this particular list of examples.

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  benchmarks. These tests can fail as dependencies are upgraded or the system
+  is under load, without actually affecting the package itself.
+
+- If your test suite takes significant time to run, **support testing
+  in parallel.** Downstreams often maintain a large number of packages,
```
```suggestion
- If your test suite takes significant time to run, support testing
  in parallel. Downstreams often maintain a large number of packages,
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  and testing them all takes a lot of time. Using pytest-xdist_ can help them
+  avoid bottlenecks.
+
+- Ideally, **support running your test suite via pytest**. pytest_ has many
```
```suggestion
- Ideally, support running your test suite via ``pytest``. pytest_ has many
```

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+  command-line arguments that are truly helpful to downstreams, such as
+  the ability to conveniently deselect tests, rerun flaky tests
+  (via pytest-rerunfailures_), add a timeout to prevent tests from hanging
+  (via pytest-timeout_) or run tests in parallel (via pytest-xdist_).
```
```suggestion
  (via pytest-timeout_) or run tests in parallel (via pytest-xdist_).
  Note that test suites don't need to be *written* with ``pytest`` to be *executed* with
  ``pytest``: ``pytest`` is able to find and execute any test cases that are compatible with
  the standard library's ``unittest`` test discovery.
```

mgorny replied:
I think this might be technically incorrect — I think pytest has some limitations / deprecations, but right now I'm not 100% sure if there were related only to nosetests leftovers, or if they included some "technically valid" unittest stuff as well.

mgorny replied:
https://docs.pytest.org/en/7.1.x/how-to/unittest.html says "almost all", so I've just added "almost" to your text.

ncoghlan replied:
Sounds good to me!

ncoghlan reviewed (commented):
Thanks for this!

*Lots* of comments inline, but many of them are just mechanical and stylistic things:

* suggested adding blank lines after the headings
* suggested changing the heading anchors to use-shorter-phrases-with-hyphens rather than repeating the titles
* suggested dropping most of the bold emphasis from the text (mostly dropped entirely, occasionally scaled back to just italics rather than bold)
* suggested rewording for anything containing the pronouns "you" or "your" (it's really hard to make second person pronouns sound non-confrontational in English, so it's simpler to avoid them entirely)
* suggested rewording anything containing the pronouns "we" or "us" or "I" (as a shared community documentation project, first person pronouns sometimes make sense as a way of referring to Python developers or project contributors in general, but we don't want to use them to refer to specific community subsets)
* took a close look at anything containing other pronouns (third party pronouns are generally fine, but the text needs to be clear on the intended reference)

There are a few more substantial comments mixed in:

* suggesting a different discussion title
* one case where I think a comment in the text isn't accurate (for projects that support typechecking, downstreams probably *should* be running a static typecheck to check for potential problems with dependency upgrades)
* a couple of cases where I suggested added a new bullet point to the rationale for a section 
* a few cases of more substantial suggested changes to the wording of various points, or adding entirely new paragraphs to the text

mgorny commented:
Thanks for your feedback.

> * suggested dropping most of the bold emphasis from the text (mostly dropped entirely, occasionally scaled back to just italics rather than bold)

Not sure about this. It's been pointed out to me that there are people who would like to skim the text rather than reading it in detail. Bold makes it easy to find the key points, especially if you've read it once and now just want to quickly consult.

> * one case where I think a comment in the text isn't accurate (for projects that support typechecking, downstreams probably should be running a static typecheck to check for potential problems with dependency upgrades)

To be honest, type checking has never caused us anything but trouble. I suppose type checking may make sense if there is a poor test coverage. However, if the test coverage is reasonably good, any real breakage affecting users should be detected through regular tests. Type checking, on the other hand, will generally report things that are "technically wrong", but do not really affect affect end users (i.e. there is no hurry to backport fixes to stable versions of a package). On top of that, MyPy is being constantly improved, so packages that passed typing tests in the past, start failing with newer MyPy versions.

Hmm, perhaps I should also mentioned the existence of "stable" channels somewhere.

mgorny commented:
> * suggested rewording for anything containing the pronouns "you" or "your" (it's really hard to make second person pronouns sound non-confrontational in English, so it's simpler to avoid them entirely)

Perhaps https://packaging.python.org/en/latest/contribute/#conventions-and-mechanics needs to be updated for that, because using "you" is not my usual style, and I've done that specifically to follow the guidelines.

mgorny reviewed (commented):
Thanks for your review. I've applied most of the suggestions, left a few with comments.

mgorny reviewed (commented):
Okay, I've attempted to address the remaining suggestions (modulo the one about Rust), taking my comments into consideration. Hope that addresses your concerns. I'm going to add one more section now, regarding release stability.

mgorny commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,442 @@
+.. _downstream-packaging:
+
+=================================
+Simplifying downstream packaging
```
Hmm, thinking about it, I'm not sure if "simplifying" is really the correct term. It suggests the discussion would be about how to change downstream packaging to make it simpler. Perhaps "helping" would be a better term?

pfmoore replied:
"Supporting"?

abravalheri commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,444 @@
+How?
+~~~~
+
+A good compromise between the needs of both parties is to provide a switch
+between using vendored and system dependencies. Ideally, if the package has
+multiple vendored dependencies, it should provide both individual switches
+for each dependency, and a general switch to control the default for them,
+e.g. via a ``USE_SYSTEM_DEPS`` environment variable.
+
+If the user requests using system dependencies, and a particular dependency
+is either missing or incompatible, the build should fail with an explanatory
+message rather than fall back to a vendored version. This gives the packager
+the opportunity to notice their mistake and a chance to consciously decide
+how to solve it.
```
This seems like a big ask, might not be viable maintenance from the perspective of the devs.

I think that in order to do this we need more substantial changes in terms of developing the infrastructure to allow it (e.g. feature from the importlib/import machinery directly).

mgorny replied:
I'm sorry but I don't understand how importlib would help when linking to system libraries.

abravalheri replied:
The impression that I had when reading the text is that it emphasizes on linking but it does mention dependencies in general. For vendored dependencies that compile a Python extension, I believe the importlib/import machinery is relevant.

If the concern is exclusive to linking, then I agree with you. Maybe the text could clarify the intention?

mgorny replied:
I focused specifically on external C/C++ dependencies, as mention on the top. I didn't want to get into details of Python dependencies — especially that vendoring Python dependencies is rather rare.

abravalheri replied:
Does it make sense to have as title `Support linking against binary system libraries`?

mgorny replied:
Perhaps. What I really wanted to convey here that there could be dependencies that aren't technically binary libraries, such as Eigen (a header-only library), or tools like CMake.

ncoghlan replied:
A potential wording improvement here might be to emphasise that this is less about folks writing these mechanisms themselves, and more about being open to PRs from downstream repackagers that add them (so the downstream redistributors aren't all having to carry a common set of patches to enable the use of system dependencies).

When submitting such PRs, the repackagers can link here to explain the purpose of the change rather than having to explain the concept from scratch themselves.

mgorny replied:
Well, I think we're already making it clear that we're willing to prepare patches for everything in the top section, so I don't think we should keep repeating it. However, the point about patches is very good, so I've added it to that section — also noting that duplication may result in inconsistent solutions.

pfmoore replied:
Reiterating my point about this being a co-operative process, I think it's worth *also* talking about what downstream repackagers can do to ease the maintenance burden of taking such a PR. Because it's often not the initial PR that's the problem, but the change in workflow, or the extra knowledge needed from the maintainer, which is the real cost.

mgorny replied:
Do you have something specific in mind?

konstin commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,445 @@
+between using vendored and system dependencies. Ideally, if the package has
+multiple vendored dependencies, it should provide both individual switches
+for each dependency, and a general switch to control the default for them,
+e.g. via a ``USE_SYSTEM_DEPS`` environment variable.
```
I would prefer something that is part of the PEP 517 interface instead of using environment variables.

There should also be some discussion about how PEP 517 build systems forward this to their native build systems, e.g. how do pyo3 users setup their `Cargo.toml` so that maturin can use this switch?

ncoghlan replied:
Improving the suggested techniques here feels like something that is best  left to a future revision, since we don't *have* a good standardised mechanism (other than process environment variables) for passing information through to build systems.

Even `pip` itself relies on environment variables to consistently apply constraints files to isolated builds.

konstin commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+Why?
+~~~~
+Some Python projects have non-Python dependencies, such as libraries written
+in C or C++. Trying to use the system versions of these dependencies
```
cryptography and pydantic are the https://github.com/pypa/packaging.python.org/issues/22 and https://github.com/pypa/packaging.python.org/issues/24 most downloaded packages on pypi ([hugovk.github.io/top-pypi-packages](https://hugovk.github.io/top-pypi-packages/)) and both use a shared library written in rust. In my experience, maturin, uv and py-spy being rust binaries are the exception. Usually, they can dynamically or statically link system dependencies written in C or C++, but the vendoring concerns apply to them too.

mgorny replied:
> cryptography and pydantic are the https://github.com/pypa/packaging.python.org/issues/22 and https://github.com/pypa/packaging.python.org/issues/24 most downloaded packages on pypi ([hugovk.github.io/top-pypi-packages](https://hugovk.github.io/top-pypi-packages/)) and both use a shared library written in rust.

I'm confused now. Are we talking of a system library written in Rust, or a library built internally?

I feel like we're splitting hairs here. I'm not saying there could be no relevant Rust dependency. What I'm saying is that adding "Rust" here is likely to add more confusion than gain, and these are just a few examples, not an exhaustive list of programming languages used to produce shared libraries.

JelleZijlstra replied:
Agree C and C++ make more sense to mention here. The libraries could also be written in Fortran or Rust but C/C++ are by far the most common in this context.

ncoghlan replied:
The Rust libraries also typically won't *have* system versions available to potentially link against, so agreed that it makes sense to omit Rust from the list of examples here.

pawamoy commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,445 @@
+
+A good idea is to **use your source distribution in the release workflow**.
+That is, build it first, then unpack it and perform all the remaining steps
+using the unpacked distribution rather than the git repostiry — run tests,
```
```suggestion
using the unpacked distribution rather than the git repository — run tests,
```

pawamoy commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,445 @@
+its documentation, and any other files that may be useful to end users, such as
+shell completions, editor support files, and so on.
+
+Some projects are concerned about increasing the size of source distribution,
```
```suggestion
Some projects are concerned about increasing the size of source distributions,
```

pawamoy commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,445 @@
+  of the *purpose* of a downstream packaging ecosystem. Helping to support that makes it
+  easier for users of those systems to access upstream projects in their preferred format.
+
+- in many cases, reliably sharing dynamic dependencies between components is a large part
+  of the *purpose* of a downstream packaging ecosystem. Helping to support that makes it
+  easier for users of those systems to access upstream projects in their preferred format.
```
Duplicated item?

mgorny replied:
Good catch, tanks. Must've accidentally got duplicated when merging changes.

pawamoy reviewed (commented):
Other general suggestions:

- write "Git" instead of "git" everywhere 
- use consistent casing in bullet points: sentence starts with capital letter, or not

Very nice document! Super interesting and helpful :)

mgorny commented:
Hopefully this version is clearer. In particular, I've tried to emphasize that we're talking about explicitly fetching stuff such as vendored dependencies, and not installing dependencies as done by a build frontend.

JelleZijlstra commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,466 @@
+distribution actually works, and that it won't accidentally install fewer files
+than the official wheels.
+
+Ideally, use the source distribution also run tests, build documentation,
```
```suggestion
Ideally, also use the source distribution to run tests, build documentation,
```

JelleZijlstra commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,466 @@
+and check for changes in them.
+
+Some projects are concerned about increasing the size of source distributions,
+or do not wish Python packaging tools to fall back to source distributions
```
I sort of feel this assumes too much. I know enough about Python packaging (I think!) to know that this is alluding to sdist files on PyPI (or other indexes), but I think that could made more explicit. It's probably worth explicitly saying something like "put your sdist on PyPI" (or other indexes), and linking terms like "sdist" to definitions.

mgorny replied:
Ah, good idea. I've gotten mixed up between assuming "source distribution" as PyPI specific term, and simultaneously using it to mean source archives in general.

JelleZijlstra commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,466 @@
+  or suffer from temporary problems that could cause the process to fail
+  or hang.
+
+- The remote resources may become temporarily or even permanently
```
Also worth adding that the external resource may change, making your build not reproducible?

mgorny replied:
Will add, thanks.

JelleZijlstra commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,466 @@
+is to facilitate more effective collaboration between upstream projects and downstream
+repackagers, not to suggest upstream projects take on tasks that downstream repackagers
+are better equipped to handle.
+Note that it is reasonable for upstream projects to leave *testing* of building with
```
Paragraph is duplicated.

mgorny replied:
Thanks. GitHub is clearly messing merging suggestions up sometimes :-(.

JelleZijlstra commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,466 @@
+the opportunity to notice their mistake and a chance to consciously decide
+how to solve it.
+
+Note that it is reasonable for upstream projects to leave *testing* of building with
```
```suggestion
It is reasonable for upstream projects to leave *testing* of building with
```
Style nit: "Note that" generally adds no meaning, prefer fewer words to say the same thing.

JelleZijlstra reviewed (commented):
Thanks, this is very helpful!

mgorny commented:
@JelleZijlstra, hope this addresses your concern. I've rewritten the paragraphs to make it clearer when I'm taking of PyPI, and when of some other hosting. I've also mentioned the possibility of installing tests.

JelleZijlstra commented on source/discussions/downstream-packaging.rst:
```diff
@@ -108,20 +108,30 @@ the files listing these dependencies (for example, ``requirements*.txt`` files)
+elsewhere, for example by attaching it to a GitHub release. Alternatively,
+large files, such as test data, can be split into separate archives.
+
+On the other hand, some projects (NumPy_, for instance) decide to install tests
```
```suggestion
On the other hand, some projects (NumPy_, for instance) decide to include tests
```

mgorny replied:
Not sure about this. It could be ambiguous over what "Python package" means here, in particular whether it doesn't just mean including them in the source distribution of the package.

mgorny replied:
Perhaps "include tests in their installed packages"?

JelleZijlstra replied:
I think that would be clearer, yes.

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+- Downstreams often **use checksums to verify the authenticity** of source files
+  on subsequent builds, which require that they remain bitwise identical over
+  time. For example, automatically generated git archives do not guarantee
+  that.
```
For this use case, there's an important property that hashes offer that checksums don't: resistance against the *intentional* generation of collisions. If we're only going to mention one of the two categories, it should be hashes.

Checksums offer the property that small bit differences in the input will change the checksum, but don't offer *any* assurance that significantly different inputs will give different checksum (e.g. some checksum algorithms won't detect full byte swaps, since they're not dependent on the order of operations). Most significantly, given a current checksum and a desired checksum, it's often possible to determine what input data to insert or append to get the desired result. Still good for checking transmission integrity (since they're usually cheaper to calculate than hashes), but not trustworthy when it comes to preventing malicious substitution of artifacts.

Hash algorithms offer much stronger collision resistance guarantees, to the point that they're considered broken if there are known techniques for producing distinct inputs that emit the same hash more efficiently than would be expected given the number of bits in the hash output. This can detect both transmission errors *and* intentional substitution.

(And after writing all that, I noted the "Outdated" tag and realised the wording had already been changed accordingly, so there's no further changes needed here)

ncoghlan commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,331 @@
+
+- Accessing remote servers poses a **privacy** issue and a potential
+  **security** issue, as it exposes information about the system building
+  the package.
```
Makes sense, nothing further to do here.

ncoghlan commented:
There are still a few inline conversations taking place (and the overall PR is still marked as a draft), so I'm not going to merge this yet, but I wanted to note that the proposed text *is* at a point now where I would be prepared to accept and publish it (we can always make further clarifications after its initial publication, after all).

mgorny commented:
Thanks. I think I've addressed all the feedback so far to the best of my ability, either by applying changes or by commenting. I see GitHub hasn't marked some of the suggestions resolved, even though they were applied — unless I've missed one, they should be marked resolved now. So yeah, unless others have comments, I think this is ready to land.

mgorny commented:
@ncoghlan, gentle ping. Can it be merged now?

mgorny commented:
Gentle ping.

mcepl commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,480 @@
+Why?
+~~~~
+
+The vast majority of downstream packagers prefer to build packages from source,
```
“Prefer” is not enough strong word here. For most distributions (all Linux ones for sure) not having source code available is the red line, which we don’t cross over.

mgorny replied:
Well, you have a point here, though I need to think how to word it better. Yes, there is a fair number of Linux distributions that distribute free packages only, but definitely **not all of them**. And then, "having source available" isn't necessarily equivalent to actually using it (I mean, there are packagers who use upstream binary packages for some reason even though they could use sources). And we are talking of Python here, and for pure Python packages "wheels" supply the "source code" just as well as source distributions.

But yes, I think it's a fair point to make that for some downstreams building from source is a must. I just don't think we can use "vast majority" anymore then.

mcepl commented on source/discussions/downstream-packaging.rst:
```diff
@@ -0,0 +1,480 @@
+Support building against system dependencies
+--------------------------------------------
+
+Why?
```
I think the order of paragraphs in the section is in the wrong order: I would first explain why dynamic linking and using system libraries is important and only then add that there might be some problems with it and static linking/vendoring may be the only solution.

mgorny replied:
I think this order makes comprehension easier. It says "you may want to do it because A, however none of A apply to downstream redistribution, and there are important reasons B to provide an option not to do it".

I think the other order of "you shouldn't do it because B, but you will do it anyway because of A, however A doesn't really apply to us, so give us an option" is more back-and-forth.

ncoghlan replied:
Given the target audience, I agree with @mgorny on this one - the current order starts by saying things the target audience (upstream developers) will typically agree with, which is then likely to put them in a more receptive mood when we explain the differences between downstream and upstream.

If we start the other way around (by covering the downstream perspective first), upstream developers will likely have already tuned out by the time we attempt to explain that we understand that "build for cross-platform portability" vs "build for platform internal consistency" drives genuinely different priorities and desirable outcomes in this space.

mcepl reviewed (commented):
Please, consider my comments for the further improvement of this text. Generally, I like it very much! Thank you!

ncoghlan reviewed (approved):
Thanks for the nudge, as checking back on the discussion here had indeed dropped off my radar.

Let's get this published, and then potentially iterate on specific subsections from there.

mgorny commented:
Thank you!
